### ISSUE TITLE
[FEA]: Add an owning stream wrapper

### ISSUE THREAD

**Issue by cliffburdick** [NVIDIA/cccl#1447]
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

General CCCL

### Is your feature request related to a problem? Please describe.

`cudaStream_t` is an opaque pointer type that has many issues that don't need to be discussed here. It would be useful to have an owning stream type that can be used across libraries so it's destroyed at the appropriate time.

### Describe the solution you'd like

Provide an API for something like a `unique_stream` and `shared_stream`.

### Describe alternatives you've considered

Making our own wrapper, but then it can't be used in other libraries.

**Comment by jrhemstad** [comment]
I agree with this. It's been something we've wanted to do for a while and just hadn't gotten around to it.

RMM has [the equivalent of a `unique_stream` ](https://github.com/rapidsai/rmm/blob/63335f23b277b02a39ce031ed4f22a576b06e6c5/include/rmm/cuda_stream.hpp#L40)that we could use as a starting point.
